src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs

```diff
@@ -17,6 +17,77 @@ public partial class Socket
 
         internal void ReplaceHandleIfNecessaryAfterFailedConnect() { /* nop on Windows */ }
 
+        public Socket(SocketInformation socketInformation)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            InitializeSockets();
+
+            SocketError errorCode = SocketPal.CreateSocket(socketInformation, out _handle,
+                ref _addressFamily, ref _socketType, ref _protocolType);
+
+            if (errorCode != SocketError.Success)
+            {
+                Debug.Assert(_handle.IsInvalid);
+                _handle = null;
+
+                if (errorCode == SocketError.InvalidArgument)
+                {
+                    throw new ArgumentException(SR.net_sockets_invalid_socketinformation, nameof(socketInformation));
+                }
+
+                // Failed to create the socket, throw.
+                throw new SocketException((int)errorCode);
+            }
+
+            if (_addressFamily != AddressFamily.InterNetwork && _addressFamily != AddressFamily.InterNetworkV6)
+            {
+                _handle.Dispose();
+                _handle = null;
+                throw new NotSupportedException(SR.net_invalidversion);
+            }
+
+            _isConnected = socketInformation.GetOption(SocketInformationOptions.Connected);
+            _willBlock = !socketInformation.GetOption(SocketInformationOptions.NonBlocking);
+            InternalSetBlocking(_willBlock);
+            _isListening = socketInformation.GetOption(SocketInformationOptions.Listening);
+
+            IPAddress tempAddress = _addressFamily == AddressFamily.InterNetwork ? IPAddress.Any : IPAddress.IPv6Any;
+            IPEndPoint ep = new IPEndPoint(tempAddress, 0);
+
+            Internals.SocketAddress socketAddress = IPEndPointExtensions.Serialize(ep);
+            errorCode = SocketPal.GetSockName(_handle, socketAddress.Buffer, ref socketAddress.InternalSize);
+            if (errorCode == SocketError.Success)
```

Comment:
And if it's not success, it's ok to do nothing?

Reply:
Framework does not throw when this fails, I left it like that for compatibility. Was it a bad approach? Is it better to be in favor of fail-fast instead?

Reply:
My preference would be to not have a connected socket constructed without a `_rightEndPoint`, as it would not get set elsewhere.

I'm not sure why `getsockname` would return an error here other than for an unbound socket. I feel like this could reasonably be:

```c#
if(errorCode == SocketError.Success)
{
   _rightEndPoint = ep.Create(socketAddress);
}
else if(errorCode == SocketError.InvalidArgument)
{
   // Socket is not yet bound.
}
else
{
   throw new SocketException((int)errorCode);
}
```

Reply:
What do we do when an "if this happens it's definitely a bug in our code" execution path gets hit?
Shall we throw? Or just `Debug.Assert()` and forget?

Reply:
I'd say assert if we are checking against our own code, throw if we are checking against external code.